packages/simulation/src/convert.ts

```diff
@@ -0,0 +1,571 @@
+export function createSymbolsToStateMap (initialTransitionTable: {[key: StateID]: [StateID, ReadSymbol][]}, numberOfNFAStates: number, symbolsPresent: Set<ReadSymbol>, curInitialState: StateID, curFinalStates: StateID[]): [{[key: StateID]: [StateID, ReadSymbol][]}, StateID, StateID[]] {
+  let symbolToStatesMap: {[key: ReadSymbol]: [StateID, StateID][]} = {}
+  const newCombinedStates: {[key: string]: {symbol: ReadSymbol, tostates: StateID[]}[]} = {}
+  const symbolsArray = Array.from(symbolsPresent)
+  for (let fromStateID = 0; fromStateID < numberOfNFAStates; fromStateID++) {
+    // If the state is not present then it has been removed and we just ignore it
+    if (initialTransitionTable[fromStateID] !== undefined) {
+      const transitions = initialTransitionTable[fromStateID]
+      for (let curElem = 0; curElem < transitions.length; curElem++) {
+        const toStateID = transitions[curElem][0]
+        const symbol = transitions[curElem][1]
+        if (symbol in symbolToStatesMap) {
+          symbolToStatesMap[symbol].push([fromStateID, toStateID])
+        } else {
+          symbolToStatesMap[symbol] = [[fromStateID, toStateID]]
+        }
+      }
+    }
+  }
+
+  // Now we will continue to create new states to represent merged states until no new states can be created
+  const keySet = new Set<string>()
+  let newKeyAdded = true
+  let tempFromState = 0
+
+  // This while loop ensures that once states are merged, their to states are merged and checked to see if they exist. If they all do, then there are no new combinations to be made and just exit, otherwise
+  // the states need to be merged
+  while (newKeyAdded) {
+    newKeyAdded = false
+    for (const symbol of Object.keys(symbolToStatesMap)) {
+      symbolToStatesMap[symbol].forEach(([curTransitionElem]) => {
+        const [curKeyCombination, combinedState] = combineStates(initialTransitionTable, symbolsArray, curTransitionElem, symbolToStatesMap, symbol)
+        if (combinedState) {
+          if (!keySet.has(curKeyCombination)) {
+            newCombinedStates[curKeyCombination] = combinedState
+            keySet.add(curKeyCombination)
+            newKeyAdded = true
+          }
+        }
+      })
+    }
+    symbolToStatesMap = {}
+    // The from state is just temporary and is not important, so use to identify the sequences
+    tempFromState = 0
+    for (const [key, value] of Object.entries(newCombinedStates)) {
+      for (let transElem = 0; transElem < value.length; transElem++) {
+        if (newCombinedStates[key][transElem].symbol in symbolToStatesMap) {
+          for (let toStateElem = 0; toStateElem < newCombinedStates[key][transElem].tostates.length; toStateElem++) {
+            symbolToStatesMap[newCombinedStates[key][transElem].symbol].push([tempFromState, newCombinedStates[key][transElem].tostates[toStateElem]])
```

Comment:
`newCombinedStates[key][transElem]` is accessed a lot, would be good to move it into a variable